src/sentry/db/models/base.py

```diff
@@ -129,3 +131,29 @@ def __model_class_prepared(sender: Any, **kwargs: Any) -> None:
 signals.pre_save.connect(__model_pre_save)
 signals.post_save.connect(__model_post_save)
 signals.class_prepared.connect(__model_class_prepared)
+
+
+class ServerModeDataError(Exception):
+    pass
+
+
+def available_on(mode: ServerComponentMode) -> Callable[[type], type]:
+    """Decorate a model class that should be active only in one mode."""
+    from . import BaseQuerySet
+
+    def decorator(decorated_class: Type[BaseModel]) -> Type[BaseModel]:
+        def queryset_override() -> BaseQuerySet:
+            raise ServerModeDataError(
+                f"{decorated_class.__name__} is available only in the {mode.value} server mode"
+            )
+
+        if not issubclass(decorated_class, BaseModel):
+            raise ValueError("`@available_on` must decorate a Model class")
```

Comment:
Once we get this merged and applied to existing tables, we should consider having a consistency test to ensure that all models now and in the future have this annotation applied.